src/main/java/com/cloudbees/jenkins/plugins/bitbucket/filesystem/BitbucketSCMFileSystem.java

```diff
@@ -0,0 +1,144 @@
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.mixin.ChangeRequestCheckoutStrategy;
+
+public class BitbucketSCMFileSystem extends SCMFileSystem {
+
+    private final String ref;
+    private final BitbucketApi api;
+
+    protected BitbucketSCMFileSystem(BitbucketApi api, String ref, SCMRevision rev) throws IOException {
+        super(rev);
+        this.ref = ref;
+        this.api = api;
+    }
+
+    @Override
+    public long lastModified() throws IOException {
+        // TODO figure out how to implement this
+        return 0L;
+    }
+
+    @NonNull
+    @Override
+    public SCMFile getRoot() {
+        return new BitbucketSCMFile(this, api, ref);
+    }
+
+    @Extension
+    public static class BuilderImpl extends SCMFileSystem.Builder {
+
+        @Override
+        public boolean supports(SCM source) {
+            return false;
+        }
+
+        @Override
+        public boolean supports(SCMSource source) {
+            return source instanceof BitbucketSCMSource;
+        }
+
+        @Override
+        public SCMFileSystem build(@NonNull Item owner, @NonNull SCM scm, @CheckForNull SCMRevision rev) {
+            return null;
+        }
+        
+        private static StandardUsernamePasswordCredentials lookupScanCredentials(@CheckForNull Item context,
+                @CheckForNull String scanCredentialsId) {
+			if (Util.fixEmpty(scanCredentialsId) == null) {
+				return null;
+			} else {
+				return CredentialsMatchers.firstOrNull(
+						CredentialsProvider.lookupCredentials(
+								StandardUsernamePasswordCredentials.class,
+								context,
+								context instanceof Queue.Task
+								? Tasks.getDefaultAuthenticationOf((Queue.Task) context)
+										: ACL.SYSTEM
+								),
+								CredentialsMatchers.withId(scanCredentialsId)
+						);
+			}
+		}
+
+        @Override
+        public SCMFileSystem build(@NonNull SCMSource source, @NonNull SCMHead head, @CheckForNull SCMRevision rev)
+                throws IOException, InterruptedException {
+            BitbucketSCMSource src = (BitbucketSCMSource) source;
+            
+            String credentialsId = src.getCredentialsId();
+            String owner = src.getRepoOwner();
+            String repository = src.getRepository();
+            String serverUrl = src.getServerUrl();
+            StandardUsernamePasswordCredentials credentials;
+            credentials = lookupScanCredentials(src.getOwner(), credentialsId);
+            
+            BitbucketApi apiClient = BitbucketApiFactory.newInstance(serverUrl, credentials, owner, repository);
+           		
+            String ref;
+            if (head instanceof BranchSCMHead) {
+                ref = head.getName();
+            } else if (head instanceof PullRequestSCMHead) {
+                PullRequestSCMHead pr = (PullRequestSCMHead) head;
+                if (!(pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) && pr.getRepository() != null) {
```

Comment:
So reading this we only support not merges?

Reply:
@casz I wasn't sure if merge strategy is going to work? I am happy to remove this check if others can confirm.

Reply:
Perhaps we can use the git client plugin to merge two files?
```touch tempfile
git merge-file file1 file3 file2

Reply:
Perhaps we can use the git client plugin to merge two files?
```
git merge-file pr.targetfile pr.targetfile pr.sourcefile
```

Reply:
Perhaps yes. For a given PR get list of files affected and then merge them by getting each file vi API. However, I would wait for how https://issues.jenkins-ci.org/browse/JENKINS-43194 gets resolved for GitHub case and then implement PR merge case for bitbucket.

Reply:
but ya it can get hairy quickly :P perhaps it is something for SCM api to solve?

Reply:
I am not too familiar with that part of code but yeah should be handled upstream.

Reply:
I would properly end up rolling my own release where we just assume PR head file is correct without trying to merge for now :|

Reply:
Current code is pretty much doing that, no?

Reply:
Then I miss read the conditional 😐

Reply:
Removing the !(pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) condition works fine for me with BitBucket Cloud and Mercurial:

Obtained Jenkinsfile from 04d3ecd41dc582ce8f0a81f6fdcd298d21b82d93+8d330304c2bf78a8f95c0db7700af37ad76b3ffe

It seems it's supported by the API

Reply:
At least when working with a single repository... I have not tried a PR from a fork

Reply:
Ok, removing the condition does not work if the PR comes from a fork:

```
ERROR: Could not do lightweight checkout, falling back to heavyweight
java.io.FileNotFoundException: URL: https://api.bitbucket.org/2.0/repositories/lpiedra/example/src/foo/Jenkinsfile
	at com.cloudbees.jenkins.plugins.bitbucket.client.BitbucketCloudApiClient.getRequestAsInputStream(BitbucketCloudApiClient.java:558)
	at com.cloudbees.jenkins.plugins.bitbucket.client.BitbucketCloudApiClient.getFileContent(BitbucketCloudApiClient.java:728)
	at com.cloudbees.jenkins.plugins.bitbucket.filesystem.BitbucketSCMFile.content(BitbucketSCMFile.java:81)
	at jenkins.scm.api.SCMFile.contentAsString(SCMFile.java:338)
	at org.jenkinsci.plugins.workflow.multibranch.SCMBinder.create(SCMBinder.java:104)
....
```

Anyway allowing merges for PR in the same repository would be still valuable

Reply:
@LuisPiedra Jenkinsfile changes from fork is not trusted revision. I think we can determine if pr from same origin but it needs to determine what revision to return. e.g. if both files are same or if merge can go thru, that is if its going to be fast forward case. Problem is, if its case where file is modified at both places, at origin and in PR - maybe give error in that case.

I think we should let this PR go without lightweight checkout for PR merge case. We can follow it up in different PR.